File: src/lib/rss.ts
import { ForumChannel, ThreadAutoArchiveDuration, Client } from "discord.js";
import type Parser from "rss-parser";
import TurndownService from "turndown";
import { parser } from "./config";
import { processedItems, saveCurrentState } from "./storage";
import type { RssSource } from "../type/types";

export async function checkRssFeeds(client: Client): Promise<void> {
  console.log(`[${new Date().toISOString()}] Checking RSS feeds...`);

  let rssConfig: Record<string, RssSource>;
  try {
    const configFile = Bun.file("./rss.toml");
    const content = await configFile.text();
    rssConfig = Bun.TOML.parse(content) as Record<string, RssSource>;
  } catch (error) {
    console.error("Failed to load rss.toml:", error);
    return;
  }

  for (const [name, config] of Object.entries(rssConfig)) {
    try {
      await processRssFeed(name, config, client);
    } catch (error) {
      console.error(`[${name}] Failed to check RSS feed:`, error);
    }
  }
}

function getValueByPath(obj: any, path: string | undefined): any {
  if (!path) return undefined;
  if (obj[path] !== undefined) return obj[path];
  return path.split(".").reduce((acc, part) => acc && acc[part], obj);
}

function getFieldValue(
  item: Parser.Item,
  config: RssSource,
  field: keyof NonNullable<RssSource["setup"]>
): any {
  const path = config.setup?.[field];
  const itemany = item as any;
  if (path) {
    return getValueByPath(itemany, path);
  }

  switch (field) {
    case "title":
      return item.title;
    case "link":
      return item.link;
    case "content":
      return item.content || item.contentSnippet;
    case "pubDate":
      return item.pubDate;
    case "author":
      return (
        itemany.creator ||
        (typeof itemany.author === "string"
          ? itemany.author
          : itemany.author?.name)
      );
    default:
      return undefined;
  }
}

async function processRssFeed(
  name: string,
  config: RssSource,
  client: Client
): Promise<void> {
  const feed = await parser.parseURL(config.rss);

  if (!processedItems.has(name)) {
    processedItems.set(name, new Set());
  }

  const processed = processedItems.get(name)!;

  // 현재 RSS 피드에 있는 항목들의 ID 수집
  const currentFeedIds = new Set(
    feed.items
      .map((item) => getFieldValue(item, config, "link") || item.guid || "")
      .filter(Boolean)
  );

  // RSS 피드에서 사라진 항목들을 처리된 목록에서 제거
  const removedItems: string[] = [];
  for (const itemId of processed) {
    if (!currentFeedIds.has(itemId)) {
      removedItems.push(itemId);
    }
  }

  if (removedItems.length > 0) {
    for (const itemId of removedItems) {
      processed.delete(itemId);
    }
    console.log(
      `[${name}] Cleaned up ${removedItems.length} old items from processed list`
    );
    saveCurrentState();
  }

  for (const item of feed.items) {
    const itemId = getFieldValue(item, config, "link") || item.guid || "";

    if (!itemId || processed.has(itemId)) continue;

    // 카테고리 필터링
    if (shouldFilterByCategory(config, item)) {
      const title = getFieldValue(item, config, "title") || "Untitled";
      console.log(
        `[${name}] Filtered by category: ${title} (${item.categories?.join(
          ", "
        )})`
      );
      processed.add(itemId);
      continue;
    }

    const title = getFieldValue(item, config, "title") || "Untitled";
    console.log(`[${name}] New item found: ${title}`);

    await postToForum(config, item, client);
    processed.add(itemId);
    saveCurrentState();
  }
}

function shouldFilterByCategory(config: RssSource, item: Parser.Item): boolean {
  if (!config.category_filter?.length) return false;

  const categories = item.categories || [];
  return categories.some((cat) => config.category_filter!.includes(cat));
}

async function postToForum(
  config: RssSource,
  item: Parser.Item,
  client: Client
): Promise<void> {
  const startTime = performance.now();

  try {
    const channel = await client.channels.fetch(config.channel);

    if (!channel || !(channel instanceof ForumChannel)) {
      console.error(
        `Channel ${config.channel} not found or is not a forum channel.`
      );
      return;
    }

    const title = (getFieldValue(item, config, "title") || "Untitled").slice(
      0,
      100
    );
    const content = buildContent(config, item);

    await channel.threads.create({
      name: title,
      autoArchiveDuration: ThreadAutoArchiveDuration.OneWeek,
      message: { content },
      appliedTags: [config.tag],
    });

    const endTime = performance.now();
    console.log(
      `[Forum] Post completed: ${title} (${(endTime - startTime).toFixed(2)}ms)`
    );
  } catch (error) {
    const endTime = performance.now();
    console.error(
      `[Forum] Post failed (${(endTime - startTime).toFixed(2)}ms):`,
      error
    );
  }
}

function buildContent(config: RssSource, item: Parser.Item): string {
  const parts: string[] = [];
  const turndownService = new TurndownService();

  const title = getFieldValue(item, config, "title") || "Untitled";
  const link = getFieldValue(item, config, "link");
  const content = getFieldValue(item, config, "content");
  const author = getFieldValue(item, config, "author");
  const authorLink = getFieldValue(item, config, "authorLink");
  const pubDate = getFieldValue(item, config, "pubDate");

  if (link) {
    parts.push(`# ${config.emoji} | [${title}](<${link}>)`);
  } else {
    parts.push(`# ${config.emoji} | ${title}`);
  }

  if (author) {
    if (authorLink) {
      parts.push(`-# 🖊️ [${author}](<${authorLink}>)`);
    } else {
      parts.push(`-# 🖊️ ${author}`);
    }
  }

  if (content) {
    const markdown = turndownService.turndown(content);
    parts.push(`\n${markdown.trim()}\n`);
  }

  if (pubDate) {
    const timestamp = Math.floor(new Date(pubDate).getTime() / 1000);
    if (!isNaN(timestamp)) {
      parts.push(`\n-# 🕐 <t:${timestamp}:f>`);
    }
  }

  return parts.join("\n") || link || "No content available.";
}

File: src/rss.rs
use crate::config::RssConfig;
use anyhow::Result;
use htmd::HtmlToMarkdown;
use feed_rs::model::{Feed, Entry};
use regex::Regex;

pub async fn fetch_feed(url: &str) -> Result<Feed> {
    let client = reqwest::Client::builder()
        .user_agent(std::env::var("RSS_USER_AGENT").unwrap_or_else(|_| "NekoRSS/1.0 (+https://github.com/imnyang/memos-rss)".to_string()))
        .build()?;
    
    let resp = client.get(url).send().await?;
    let status = resp.status();
    let content = resp.bytes().await?;
    
    match feed_rs::parser::parse(&content[..]) {
        Ok(feed) => Ok(feed),
        Err(e) => {
            let body_preview = String::from_utf8_lossy(&content).chars().take(200).collect::<String>();
            anyhow::bail!("Failed to parse feed (Status: {}): {} | Body preview: {}", status, e, body_preview);
        }
    }
}

pub fn should_include_by_link(link: Option<&String>, link_filters: Option<&Vec<String>>) -> bool {
    match (link, link_filters) {
        (Some(link_url), Some(filters)) => {
            // 모든 필터 중 하나라도 매치되면 포함
            filters.iter().any(|filter| {
                if let Ok(regex) = Regex::new(filter) {
                    regex.is_match(link_url)
                } else {
                    false
                }
            })
        }
        (_, None) => true, // 필터가 없으면 모두 포함
        (None, Some(_)) => false, // 필터가 있는데 link가 없으면 제외
    }
}

pub fn extract_image_url(item: &Entry) -> Option<String> {
    // 첫 번째로 media:content에서 이미지 찾기
    item.media
        .iter()
        .find(|m| m.content.iter().any(|c| c.medium.as_deref() == Some("image")))
        .and_then(|m| m.content.first())
        .and_then(|c| c.url.clone())
        .map(|url| url.to_string())
}

pub fn get_field_value(
    item: &Entry,
    config: &RssConfig,
    field: &str,
) -> Option<String> {
    let setup = config.setup.as_ref();
    let path = match field {
        "title" => setup.and_then(|s| s.title.as_ref()),
        "link" => setup.and_then(|s| s.link.as_ref()),
        "content" => setup.and_then(|s| s.content.as_ref()),
        "author" => setup.and_then(|s| s.author.as_ref()),
        "authorLink" => setup.and_then(|s| s.author_link.as_ref()),
        "pubDate" => setup.and_then(|s| s.pub_date.as_ref()),
        _ => None,
    };

    if let Some(_p) = path {
        match _p.as_str() {
            "title" => item.title.as_ref().map(|t| t.content.clone()),
            "link" => item.links.first().map(|l| l.href.clone()),
            "description" | "summary" => item.summary.as_ref().map(|s| s.content.clone()),
            "content" => item.content.as_ref().and_then(|c| c.body.clone()),
            "published" | "pubDate" => item.published.map(|d| d.to_rfc3339()),
            "dc:creator" | "author.name" => item.authors.first().map(|a| a.name.clone()),
            "author.uri" => item.authors.first().and_then(|a| a.uri.clone()),
            _ => None,
        }
    } else {
        match field {
            "title" => item.title.as_ref().map(|t| t.content.clone()),
            "link" => item.links.first().map(|l| l.href.clone()),
            "content" => item.content.as_ref().and_then(|c| c.body.clone())
                .or_else(|| item.summary.as_ref().map(|s| s.content.clone())),
            "author" => item.authors.first().map(|a| a.name.clone()),
            "pubDate" => item.published.map(|d| d.to_rfc3339()),
            _ => None,
        }
    }
}

pub fn build_content(config: &RssConfig, item: &Entry) -> String {
    let mut parts = Vec::new();
    let ht = HtmlToMarkdown::new();

    let title = get_field_value(item, config, "title").unwrap_or_else(|| "Untitled".to_string());
    let link = get_field_value(item, config, "link");
    let content = get_field_value(item, config, "content");
    let author = get_field_value(item, config, "author");
    let author_link = get_field_value(item, config, "authorLink");
    let pub_date = item.published;

    if let Some(l) = link.as_ref() {
        parts.push(format!("# {} | [{}](<{}>)", config.emoji, title, l));
    } else {
        parts.push(format!("# {} | {}", config.emoji, title));
    }

    if let Some(a) = author {
        if let Some(al) = author_link {
            parts.push(format!("-# 🖊️ [{}](<{}>)", a, al));
        } else {
            parts.push(format!("-# 🖊️ {}", a));
        }
    }

    if let Some(c) = content {
        let markdown = ht.convert(&c).unwrap_or_else(|_| c);
        parts.push(format!("\n{}\n", markdown.trim()));
    }

    let timestamp = pub_date
        .map(|pd| pd.timestamp())
        .unwrap_or_else(|| {
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs() as i64
        });
    parts.push(format!("\n-# 🕐 <t:{}:f>", timestamp));

    let joined = parts.join("\n");
    if joined.is_empty() {
        link.unwrap_or_else(|| "No content available.".to_string())
    } else {
        joined
    }
}
